File: mobile/ios/OpenU60/Features/DeviceInfo/DeviceInfoViewModel.swift
import SwiftUI

@Observable
@MainActor
final class DeviceInfoViewModel {
    var identity: DeviceIdentity = .empty
    var operatorInfo: OperatorInfo = .empty
    var isLoading: Bool = false
    var error: String?

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        error = nil

        async let simTask = fetchSIMInfo()
        async let imeiTask = fetchIMEI()
        async let wanTask = fetchWANStatus()
        async let wan6Task = fetchWAN6Status()
        async let lanTask = fetchLANStatus()
        async let signalTask = fetchSignalInfo()

        let (simInfo, imeiData, wanStatus, wan6Status, lanStatus, signalInfo) =
            await (simTask, imeiTask, wanTask, wan6Task, lanTask, signalTask)

        identity = DeviceParser.parseIdentity(
            simInfo: simInfo ?? [:],
            imeiData: imeiData ?? [:],
            wanStatus: wanStatus ?? [:],
            wan6Status: wan6Status ?? [:],
            lanStatus: lanStatus ?? [:]
        )
        if let signalInfo {
            operatorInfo = signalInfo
        }
        isLoading = false
    }

    private func fetchSIMInfo() async -> [String: Any]? {
        do {
            return try await client.getJSON("/api/sim/info")
        } catch {
            self.error = error.localizedDescription
            return nil
        }
    }

    private func fetchIMEI() async -> [String: Any]? {
        try? await client.getJSON("/api/sim/imei")
    }

    private func fetchWANStatus() async -> [String: Any]? {
        try? await client.getJSON("/api/network/wan")
    }

    private func fetchWAN6Status() async -> [String: Any]? {
        try? await client.getJSON("/api/network/wan6")
    }

    private func fetchLANStatus() async -> [String: Any]? {
        try? await client.getJSON("/api/network/lan-status")
    }

    private func fetchSignalInfo() async -> OperatorInfo? {
        do {
            let data = try await client.getJSON("/api/network/signal")
            let (_, _, _, opInfo) = SignalParser.parseNetInfo(data)
            return opInfo
        } catch { return nil }
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/deviceinfo/DeviceInfoViewModel.kt
package com.openu60.feature.deviceinfo

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.DeviceIdentity
import com.openu60.core.model.DeviceParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DeviceInfoState(
    val identity: DeviceIdentity = DeviceIdentity.empty,
    val isLoading: Boolean = false,
    val error: String? = null,
)

@HiltViewModel
class DeviceInfoViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(DeviceInfoState())
    val state: StateFlow<DeviceInfoState> = _state.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null)
            try {
                coroutineScope {
                    val simJob = async { try { agentClient.getJSON("/api/sim/info") } catch (_: Exception) { emptyMap() } }
                    val imeiJob = async { try { agentClient.getJSON("/api/device/imei") } catch (_: Exception) { emptyMap() } }
                    val wanJob = async { try { agentClient.getJSON("/api/network/wan") } catch (_: Exception) { emptyMap() } }
                    val wan6Job = async { try { agentClient.getJSON("/api/network/wan6") } catch (_: Exception) { emptyMap() } }
                    val lanJob = async { try { agentClient.getJSON("/api/network/lan") } catch (_: Exception) { emptyMap() } }

                    val identity = DeviceParser.parseIdentity(
                        simInfo = simJob.await(),
                        imeiData = imeiJob.await(),
                        wanStatus = wanJob.await(),
                        wan6Status = wan6Job.await(),
                        lanStatus = lanJob.await(),
                    )
                    _state.value = _state.value.copy(identity = identity, isLoading = false)
                }
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh()
                else _state.value = _state.value.copy(isLoading = false, error = "Session expired")
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false, error = e.message)
            }
        }
    }
}
